File: Output/output_TY/cpp/Hard/1814-remove-invalid-parentheses.cpp
/**https://leetcode.com/problems/remove-invalid-parentheses/ */
//Given a string s that contains parentheses and letters, remove the minimum number of invalid parentheses to make the input string valid.
//Return a list of unique strings that are valid with the minimum number of removals. You may return the answer in any order.
// 
//Example 1:
//Input: s = "()())()"
//Output: ["(())()","()()()"]
//Example 2:
//Input: s = "(a)())()"
//Output: ["(a())()","(a)()()"]
//Example 3:
//Input: s = ")("
//Output: [""]
// 
//Constraints:
//	1 <= s.length <= 25
//	s consists of lowercase English letters and parentheses '(' and ')'.
//	There will be at most 20 parentheses in s.
class Solution {
public:
    vector<string> removeInvalidParentheses(string s) {
        int left = 0, right = 0;
        for(auto c : s){
            if(c == '('){
                ++left;
            }
            else if(c == ')'){
                if(left){
                    --left;
                }
                else{
                    ++right;
                }
            }
        }

        unordered_set<string> result;
        dfs(s, 0, left, right, result);
        return vector<string>(result.begin(), result.end());
    }

private:
    void dfs(string& s, int start, int left, int right, unordered_set<string>&result){
        if(left == 0 && right == 0){
            if(isValid(s)){
                result.insert(s);
            }
        }
        else{
            for(int i = start; i < s.size(); ++i){
                if(i != start && s[i] == s[i - 1]){
                    continue;
                }
                if(left > 0 && s[i] == '('){
                    swap(s[i], s[start]);
                    dfs(s, start + 1, left - 1, right, result);
                    swap(s[i], s[start]);
                }
               if(right > 0 && s[i] == ')')
                {
                    swap(s[i], s[start]);
                    dfs(s, start + 1, left, right - 1, result);
                    swap(s[i], s[start]);
                    return;
                }
            }
        }
    }
};

File: Output/output_TY/javascript/Hard/1814-remove-invalid-parentheses.js
/**https://leetcode.com/problems/remove-invalid-parentheses/ */
//Given a string s that contains parentheses and letters, remove the minimum number of invalid parentheses to make the input string valid.
//Return a list of unique strings that are valid with the minimum number of removals. You may return the answer in any order.
// 
//Example 1:
//Input: s = "()())()"
//Output: ["(())()","()()()"]
//Example 2:
//Input: s = "(a)())()"
//Output: ["(a())()","(a)()()"]
//Example 3:
//Input: s = ")("
//Output: [""]
// 
//Constraints:
//	1 <= s.length <= 25
//	s consists of lowercase English letters and parentheses '(' and ')'.
//	There will be at most 20 parentheses in s.
/**
 * @param {string} s
 * @return {string[]}
 */
var removeInvalidParentheses = function(s) {
    let result = [];
    let queue = [s];
    let visited = new Set();
    let found = false;
    while(queue.length){
        let current = queue.shift();
        if(isValid(current)){
            result.push(current);
            found = true;
        }
        if(found) continue;
        for(let i = 0; i < current.length; i++){
            if(current[i] !== '(' && current[i] !== ')') continue;
            let newString = current.slice(0,i) + current.slice(i+1);
            if(!visited.has(newString)){
                queue.push(newString);
            }
        }
    }
    return result;
};